File: iact-check-core/src/main/kotlin/ch/iact/iactcheck/service/DocumentService.kt
package ch.iact.iactcheck.service

import ch.iact.iactcheck.controller.exception.CheckNotFoundException
import ch.iact.iactcheck.controller.exception.DocumentFileNotFoundException
import ch.iact.iactcheck.controller.exception.DocumentGroupNotFoundException
import ch.iact.iactcheck.controller.exception.DocumentNotFoundException
import ch.iact.iactcheck.domain.model.Document
import ch.iact.iactcheck.domain.model.DocumentFile
import ch.iact.iactcheck.domain.model.DocumentGroup
import ch.iact.iactcheck.domain.repository.CheckRepository
import ch.iact.iactcheck.domain.repository.DocumentFileRepository
import ch.iact.iactcheck.domain.repository.DocumentGroupRepository
import ch.iact.iactcheck.domain.repository.DocumentRepository
import ch.iact.iactcheck.dto.DocumentDTO
import ch.iact.iactcheck.dto.DocumentGroupDTO
import ch.iact.iactcheck.service.converter.DocumentConverter
import org.springframework.http.MediaType
import org.springframework.stereotype.Service

@Service
class DocumentService(
    private val documentGroupRepository: DocumentGroupRepository,
    private val documentFileRepository: DocumentFileRepository,
    private val documentRepository: DocumentRepository,
    private val checkRepository: CheckRepository
) {

    fun createDocumentGroup(documentGroupDTO: DocumentGroupDTO): DocumentGroupDTO {
        val check = checkRepository
            .findById(documentGroupDTO.checkId)
            .orElseThrow { throw CheckNotFoundException() }

        val documentGroup = DocumentGroup(
            id = -1,
            name = documentGroupDTO.name,
            backgroundColour = documentGroupDTO.backgroundColour,
            check = check,
            documents = ArrayList()
        )
        return DocumentConverter.map(documentGroupRepository.save(documentGroup))
    }

    fun createDocumentForDocumentGroup(documentGroupId: Long, documentDTO: DocumentDTO): DocumentDTO {
        val documentGroup = documentGroupRepository
            .findById(documentGroupId)
            .orElseThrow { throw DocumentGroupNotFoundException() }

        val document = Document(
            id = -1,
            title = documentDTO.title,
            position = documentGroup.documents.size,
            mediaType = documentDTO.mediaType,
            documentGroup = documentGroup
        )

        return DocumentConverter.map(documentRepository.save(document))
    }

    fun getDocumentGroupById(documentGroupId: Long): DocumentGroupDTO {
        return DocumentConverter.map(
            documentGroupRepository
                .findById(documentGroupId)
                .orElseThrow { throw DocumentGroupNotFoundException() }
        )
    }

    fun getDocumentById(documentId: Long): DocumentDTO {
        return DocumentConverter.map(
            documentRepository
                .findById(documentId)
                .orElseThrow { throw DocumentNotFoundException() }
        )
    }

    fun getFileByDocumentId(documentId: Long): Pair<MediaType, ByteArray> {
        val documentFile = documentFileRepository
            .findById(documentId)
            .orElseThrow { throw DocumentFileNotFoundException() }

        return Pair(MediaType.valueOf(documentFile.document.mediaType), documentFile.file)
    }

    fun getDocumentGroupsByCheckId(checkId: Long): List<DocumentGroupDTO> {
        return documentGroupRepository
            .findAllByCheckId(checkId)
            .map(DocumentConverter::map)
    }

    fun uploadFileForDocument(documentId: Long, file: ByteArray) {
        val documentFile = if (documentFileRepository.existsById(documentId)) {
            documentFileRepository
                .findById(documentId)
                .orElseThrow { throw DocumentFileNotFoundException() }
                .copy(file = file)
        } else {
            val document = documentRepository
                .findById(documentId)
                .orElseThrow { throw DocumentNotFoundException() }

            DocumentFile(file = file, document = document)
        }

        documentFileRepository.save(documentFile)
    }

    fun updateDocumentGroupById(documentGroupId: Long, documentGroupDTO: DocumentGroupDTO): DocumentGroupDTO {
        var documentGroup = documentGroupRepository
            .findById(documentGroupId)
            .orElseThrow { throw DocumentGroupNotFoundException() }

        documentGroup = documentGroup.copy(
            name = documentGroupDTO.name,
            backgroundColour = documentGroupDTO.backgroundColour,
        )

        return DocumentConverter.map(documentGroupRepository.save(documentGroup))
    }

    fun updateDocuments(documents: List<DocumentDTO>): List<DocumentDTO> {
        return documents.map { updateDocumentById(it.id, it) }
    }

    fun updateDocumentById(documentId: Long, documentDTO: DocumentDTO): DocumentDTO {
        val document = documentRepository
            .findById(documentId)
            .orElseThrow { throw DocumentNotFoundException() }
            .copy(
                title = documentDTO.title,
                mediaType = documentDTO.mediaType,
                position = documentDTO.position
            )

        return DocumentConverter.map(documentRepository.save(document))
    }

    fun deleteDocumentGroupById(documentGroupId: Long) {
        val documentsGroup = getDocumentGroupById(documentGroupId)
        documentsGroup.documents.map(DocumentDTO::id).forEach(this::deleteDocumentById)
        documentGroupRepository.deleteById(documentGroupId)
    }

    fun deleteDocumentById(documentId: Long) {
        documentFileRepository.deleteById(documentId)
        documentRepository.deleteById(documentId)
    }
}

File: iact-check-webapp/src/main/resources/webapp/src/app/shared/services/document.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { DocumentGroupDTO } from '../dtos/document-group-dto';
import { CORE_URL } from '../../app.config';
import { DocumentDTO } from '../dtos/document-dto';

@Injectable({
  providedIn: 'root',
})
export class DocumentService {
  constructor(private readonly httpClient: HttpClient) {}

  public createDocumentGroup(documentGroupDTO: DocumentGroupDTO): Observable<DocumentGroupDTO> {
    const body = JSON.stringify(documentGroupDTO);
    return this.httpClient.post<DocumentGroupDTO>(`${CORE_URL}/api/admin/document-groups`, body);
  }

  public createDocumentForDocumentGroup(documentGroupId: number, documentDTO: DocumentDTO): Observable<DocumentDTO> {
    const body = JSON.stringify(documentDTO);
    return this.httpClient.post<DocumentDTO>(
      `${CORE_URL}/api/admin/document-groups/${documentGroupId}/documents`,
      body
    );
  }

  public getDocumentGroupById(documentGroupId: number): Observable<DocumentGroupDTO> {
    return this.httpClient.get<DocumentGroupDTO>(`${CORE_URL}/api/document-groups/${documentGroupId}`);
  }

  public getDocumentById(documentId: number): Observable<DocumentDTO> {
    return this.httpClient.get<DocumentDTO>(`${CORE_URL}/api/admin/document-groups/documents/${documentId}`);
  }

  public getFileByDocumentId(documentId: number): Observable<Blob> {
    return this.httpClient.get(CORE_URL + '/api/documents/' + documentId + '/file', {
      responseType: 'blob',
    });
  }

  public getDocumentGroupsByCheckId(checkId: number): Observable<DocumentGroupDTO[]> {
    return this.httpClient.get<DocumentGroupDTO[]>(`${CORE_URL}/api/checks/${checkId}/document-groups`);
  }

  public updateDocumentGroupById(
    documentGroupId: number,
    documentGroupDTO: DocumentGroupDTO
  ): Observable<DocumentGroupDTO> {
    const body = JSON.stringify(documentGroupDTO);
    return this.httpClient.put<DocumentGroupDTO>(`${CORE_URL}/api/admin/document-groups/${documentGroupId}`, body);
  }

  public updateDocumentById(documentId: number, documentDTO: DocumentDTO): Observable<DocumentDTO> {
    const body = JSON.stringify(documentDTO);
    return this.httpClient.put<DocumentDTO>(`${CORE_URL}/api/admin/document-groups/documents/${documentId}`, body);
  }

  public updateDocuments(documents: DocumentDTO[]): Observable<DocumentDTO[]> {
    const body = JSON.stringify(documents);
    return this.httpClient.put<DocumentDTO[]>(`${CORE_URL}/api/admin/document-groups/documents`, body);
  }

  public uploadFileForDocument(documentId: number, file: File): Observable<void> {
    const body: FormData = new FormData();
    body.append('file', file, file.name);
    return this.httpClient.put<void>(CORE_URL + '/api/admin/document-groups/documents/' + documentId + '/assets', body);
  }

  public deleteDocumentGroupById(documentGroupId: number): Observable<void> {
    return this.httpClient.delete<void>(`${CORE_URL}/api/admin/document-groups/${documentGroupId}`);
  }

  public deleteDocumentById(documentId: number): Observable<void> {
    return this.httpClient.delete<void>(`${CORE_URL}/api/admin/document-groups/documents/${documentId}`);
  }
}
